Transparent, composited canvas is rendered black when layer is changed to render layer backing sharing

kkinnunen-apple opened the pull request:
#### dbef4cf540cf8c439d4ff62c949da5f3d225636f
<pre>
Transparent, composited canvas is rendered black when layer is changed to render layer backing sharing
<a href="https://bugs.webkit.org/show_bug.cgi?id=258927">https://bugs.webkit.org/show_bug.cgi?id=258927</a>
rdar://110088034

Reviewed by Simon Fraser.

Explicitly composited canvas element might switch to being
composited to other RenderLayerBacking. This left the
RenderLayer::m_shouldPaintUsingCompositeCopy unreset.

Instead, store the flag to RenderLayerBacking.

* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::paint):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/rendering/RenderHTMLCanvas.cpp:
(WebCore::RenderHTMLCanvas::paintReplaced):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::shouldPaintUsingCompositeCopy const): Deleted.
(WebCore::RenderLayer::setShouldPaintUsingCompositeCopy): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
(WebCore::RenderLayerBacking::paintIntoLayer):
* Source/WebCore/rendering/RenderLayerBacking.h:

Canonical link: <a href="https://commits.webkit.org/265837@main">https://commits.webkit.org/265837@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e43669bec6ea59212c23149e5ea152de56a19440

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12032 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12379 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12681 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13778 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11612 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12051 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14794 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12392 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14308 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12196 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13009 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10182 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14192 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10297 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10919 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18045 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11377 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11079 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14252 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11566 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9521 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10781 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2943 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15105 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11418 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayerBacking.h:
```diff
@@ -289,6 +289,9 @@ class RenderLayerBacking final : public GraphicsLayerClient {
     WEBCORE_EXPORT void setIsTrackingDisplayListReplay(bool);
     WEBCORE_EXPORT String replayDisplayListAsText(OptionSet<DisplayList::AsTextFlag>) const;
 
+#if ENABLE(CSS_COMPOSITING)
```
No need to wrap this in `#if ENABLE(CSS_COMPOSITING)`, and using that around the existing code was wrong; `ENABLE(CSS_COMPOSITING)` is about web-exposed CSS Compositing and Blending.

smfr commented on Source/WebCore/rendering/RenderLayerBacking.h:
```diff
@@ -456,6 +459,9 @@ class RenderLayerBacking final : public GraphicsLayerClient {
 #if ENABLE(ASYNC_SCROLLING)
     bool m_needsEventRegionUpdate { true };
 #endif
+#if ENABLE(CSS_COMPOSITING)
```
Ditto

smfr commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -3443,6 +3443,9 @@ void RenderLayerBacking::paintIntoLayer(const GraphicsLayer* graphicsLayer, Grap
     }
+    GraphicsContextStateSaver saver(context, m_shouldPaintUsingCompositeCopy);
+    if (m_shouldPaintUsingCompositeCopy)
+        context.setCompositeOperation(CompositeOperator::Copy);
```
It feels a bit weird to do this here. Sure, we know that the canvas is the only thing painted into this layer, but it looks weird to have this cover the scope of all the painting. Maybe it will do something odd for event regions, for example.

smfr replied:
I'd feel OK if this was moved to before `layer.paintLayerContents()`.

kkinnunen-apple committed:
Transparent, composited canvas is rendered black when layer is changed to render layer backing sharing
https://bugs.webkit.org/show_bug.cgi?id=258927
rdar://110088034

Reviewed by Simon Fraser.

Explicitly composited canvas element might switch to being
composited to other RenderLayerBacking. This left the
RenderLayer::m_shouldPaintUsingCompositeCopy unreset.

Instead, store the flag to RenderLayerBacking.

* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::paint):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/rendering/RenderHTMLCanvas.cpp:
(WebCore::RenderHTMLCanvas::paintReplaced):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::shouldPaintUsingCompositeCopy const): Deleted.
(WebCore::RenderLayer::setShouldPaintUsingCompositeCopy): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
(WebCore::RenderLayerBacking::paintIntoLayer):
* Source/WebCore/rendering/RenderLayerBacking.h:

Canonical link: https://commits.webkit.org/265837@main
